### ISSUE TITLE
Incompatibility with latest version of Dask because of `dask-expr` depending on `pandas>=2`

### ISSUE THREAD

**Issue by astrojuanlu** [kedro-org/kedro-plugins#609]
https://github.com/dask/dask/commit/a6eccb2b7cad117b2f338142391e500cecc1ef0f

```
uv pip install "kedro-datasets[test,docs] @ ." "dask-expr"
  × No solution found when resolving dependencies:
  ╰─▶ Because only the following versions of dask-expr are available:
          dask-expr==0.1.0
          dask-expr==0.1.1
          dask-expr==0.1.2
          dask-expr==0.1.3
          dask-expr==0.1.4
          dask-expr==0.1.5
          dask-expr==0.1.6
          dask-expr==0.1.7
          dask-expr==0.1.8
          dask-expr==0.1.9
          dask-expr==0.1.10
          dask-expr==0.1.11
          dask-expr==0.1.12
          dask-expr==0.2.0
          dask-expr==0.2.1
          dask-expr==0.2.2
          dask-expr==0.2.3
          dask-expr==0.2.4
          dask-expr==0.2.5
          dask-expr==0.2.6
          dask-expr==0.2.7
          dask-expr==0.2.8
          dask-expr==0.2.9
          dask-expr==0.3.0
          dask-expr==0.3.1
          dask-expr==0.3.2
          dask-expr==0.3.3
          dask-expr==0.3.4
          dask-expr==0.3.5
          dask-expr==0.4.0
          dask-expr==0.4.1
          dask-expr==0.4.2
          dask-expr==0.5.0
          dask-expr==0.5.1
          dask-expr==0.5.2
          dask-expr==0.5.3
          dask-expr==1.0
          dask-expr==1.0.1
      and any of:
          dask-expr<=0.4.2
          dask-expr>=0.5.1
      depends on pandas>=2, we can conclude that any of:
          dask-expr<0.5.0
          dask-expr>0.5.0
      depends on pandas>=2.
      And because dask-expr==0.5.0 is unusable because it was yanked (reason: Wrong Dask Version Pin), we can conclude that all versions of
      dask-expr depend on pandas>=2.
      And because kedro-datasets[test]==2.1.0 depends on pandas>=1.3,<2.dev0 and only kedro-datasets[test]==2.1.0 is available, we can conclude
      that all versions of dask-expr and all versions of kedro-datasets[test] are incompatible.
      And because you require kedro-datasets[test] and you require dask-expr, we can conclude that the requirements are unsatisfiable.
```

**Comment by astrojuanlu** [comment]
Apparently we're pinning pandas to <2 because of compatibility with older versions of Spark, see #297

**Comment by datajoely** [comment]
For reference:

| Spark Version | Release Date   |
|---------------|----------------|
| 3.5.1         | February 23, 2024 |
| 3.3.4         | December 16, 2023 |
| 3.4.2         | November 30, 2023 |
| 3.5.0         | September 13, 2023 |
| 3.3.3         | August 21, 2023 |
| 3.4.1         | June 23, 2023 |
| 3.4.0         | April 13, 2023 |
| 3.2.4         | April 13, 2023 |
| 3.3.2         | February 17, 2023 |
| 3.2.3         | November 28, 2022 |
| 3.3.1         | October 25, 2022 |
| 3.2.2         | July 17, 2022 |
| 3.3.0         | June 16, 2022 |
| 3.1.3         | February 18, 2022 |
| 3.2.1         | January 26, 2022 |
| 3.2.0         | October 13, 2021 |
| 3.0.3         | June 23, 2021 |
| 3.1.2         | June 1, 2021 |
| 2.4.8         | May 17, 2021 |
| 3.1.1         | March 2, 2021 |
| 3.0.2         | February 19, 2021 |
| 2.4.7         | September 12, 2020 |
| 3.0.1         | September 8, 2020 |
| 3.0.0         | June 18, 2020 |
| 2.4.6         | June 5, 2020 |
| 2.4.5         | February 8, 2020 |
| 2.3.4         | September 9, 2019 |
| 2.4.4         | September 1, 2019 |
| 2.4.3         | May 8, 2019 |
| 2.4.2         | April 23, 2019 |
| 2.4.1         | March 31, 2019 |
| 2.3.3         | February 15, 2019 |
| 2.2.3         | January 11, 2019 |
| 2.4.0         | November 2, 2018 |
| 2.3.2         | September 24, 2018 |
| 2.2.2         | July 2, 2018 |
| 2.1.3         | June 29, 2018 |
| 2.3.1         | June 8, 2018 |
| 2.3.0         | February 28, 2018 |
| 2.2.1         | December 1, 2017 |
| 2.1.2         | October 9, 2017 |
| 2.2.0         | July 11, 2017 |



| Pandas Version | Release Date   |
|----------------|----------------|
| 2.2.1          | February 23, 2024 |
| 2.2.0          | January 20, 2024 |
| 2.2.0rc0       | December 22, 2023 |
| 2.1.4          | December 8, 2023 |
| 2.1.3          | November 10, 2023 |
| 2.1.2          | October 26, 2023 |
| 2.1.1          | September 20, 2023 |
| 2.1.0          | August 30, 2023 |
| 2.1.0rc0       | August 11, 2023 |
| 2.0.3          | June 28, 2023 |
| 2.0.2          | May 29, 2023 |
| 2.0.1          | April 24, 2023 |
| 2.0.0          | April 3, 2023 |
| 2.0.0rc1       | March 16, 2023 |
| 2.0.0rc0       | February 20, 2023 |
| 1.5.3          | January 19, 2023 |
| 1.5.2          | November 23, 2022 |
| 1.5.1          | October 19, 2022 |
| 1.5.0          | September 19, 2022 |
| 1.4.4          | August 31, 2022 |
| 1.5.0rc0       | August 24, 2022 |
| 1.4.3          | June 23, 2022 |
| 1.4.2          | April 2, 2022 |
| 1.4.1          | February 12, 2022 |
| 1.4.0          | January 22, 2022 |
| 1.4.0rc0       | January 6, 2022 |
| 1.3.5          | December 12, 2021 |
| 1.3.4          | October 17, 2021 |
| 1.3.3          | September 12, 2021 |
| 1.3.2          | August 15, 2021 |
| 1.3.1          | July 25, 2021 |
| 1.3.0          | July 2, 2021 |
| 1.2.5          | June 22, 2021 |
| 1.2.4          | April 12, 2021 |
| 1.2.3          | March 2, 2021 |
| 1.2.2          | February 9, 2021 |
| 1.2.1          | January 20, 2021 |
| 1.2.0          | December 26, 2020 |
| 1.1.5          | December 7, 2020 |
| 1.1.4          | October 30, 2020 |

**Comment by datajoely** [comment]
Personally I think it's reasonable for us to drop Spark <3

**Comment by deepyaman** [comment]
> Personally I think it's reasonable for us to drop Spark <3

I misread this as you advocating dropping Spark, with love. Got prematurely excited there for a minute. :joy:

**Comment by datajoely** [comment]
I mean I'm all for one `IbisDataset` to rule them all
<img width="1074" alt="image" src="https://github.com/kedro-org/kedro-plugins/assets/35801847/e8f8713e-c068-4c5e-a991-a983cc648d99">

**Comment by astrojuanlu** [comment]
This is now preventing us from addressing https://github.com/dask/dask/issues/11038 and is blocking all PRs on `kedro-datasets`. Switching to an older version of Dask won't help.

**Comment by astrojuanlu** [comment]
(Dask issue spotted by @merelcht )

**Comment by noklam** [comment]
What's preventing us to move our test to support pandas 2.0?

**Comment by astrojuanlu** [comment]
The reason why this only appeared now:

- Nightly checks 16 hours ago still used Python 3.11.8 https://github.com/kedro-org/kedro-plugins/actions/runs/8608244106/job/23590256301#step:3:12
- Recently the GHA images switched to Python 3.11.9

**Comment by lrcouto** [comment]
@astrojuanlu I pushed a draft changing the version requirements [here](https://github.com/kedro-org/kedro-plugins/pull/643) but the same test still fails. Is there anything else that needs to be done?

**Comment by astrojuanlu** [comment]
Lifting the upper version cap for Dask! Pushed a commit to #643

**Comment by ElenaKhaustova** [comment]
Solved in #643
